compiler/src/iree/compiler/Codegen/Common/LinalgOpInfo.cpp

```diff
@@ -57,51 +56,84 @@ static bool isTransposeMap(AffineMap map) {
   return false;
 }
 
+/// The default filter passes all op operands.
+static bool defaultTransposeMapFilter(AffineMap map) { return true; }
+
+LinalgOpInfo::LinalgOpInfo(linalg::LinalgOp linalgOp)
+    : transposeMapFilter(defaultTransposeMapFilter) {
+  computeInfo(linalgOp);
+}
+LinalgOpInfo::LinalgOpInfo(linalg::LinalgOp linalgOp,
+                           TransposeMapFilter transposeMapFilter)
+    : transposeMapFilter(transposeMapFilter) {
+  computeInfo(linalgOp);
+}
+
 /// Returns true if a LinalgOp implements a transpose.
 // TODO(dcaballe):
 //   * Consider transpose + reductions.
 //   * Consider input and output transposes.
-static bool isTransposeLinalgOp(linalg::LinalgOp linalgOp) {
+static SmallVector<OpOperand *> computeTransposeInfo(
+    LinalgOp linalgOp, TransposeMapFilter transposeMapFilter) {
+  SmallVector<OpOperand *> transposeOperands;
+
   // Reductions are not supported.
   if (linalgOp.getNumReductionLoops() > 0) {
-    return false;
+    return transposeOperands;
   }
 
   // Multiple outputs are not supported yet.
   if (linalgOp.getNumOutputs() != 1) {
-    return false;
+    return transposeOperands;
   }
 
   // Inverse map to use transfer op permutation logic.
   AffineMap outputInversedMap = inversePermutation(
       linalgOp.getTiedIndexingMap(linalgOp.getOutputOperand(0)));
+
   SmallVector<AffineMap> inputInversedMaps;
   for (OpOperand *linalgOperand : linalgOp.getInputOperands()) {
     auto map = linalgOp.getTiedIndexingMap(linalgOperand);
     if (!map.isProjectedPermutation(/*allowZeroInResults=*/true)) {
-      return false;
+      return transposeOperands;
+    }
+    AffineMap inverseMap = inverseAndBroadcastProjectedPermutation(map);
+    if (isTransposeMap(inverseMap) && transposeMapFilter(inverseMap)) {
+      transposeOperands.push_back(linalgOperand);
     }
-    inputInversedMaps.push_back(inverseAndBroadcastProjectedPermutation(map));
   }
 
-  bool isInputTransposed = llvm::any_of(
-      inputInversedMaps, [](AffineMap map) { return isTransposeMap(map); });
-  bool isOutputTransposed = isTransposeMap(outputInversedMap);
-
-  return isInputTransposed || isOutputTransposed;
-}
+  if (isTransposeMap(outputInversedMap) &&
+      transposeMapFilter(outputInversedMap)) {
+    transposeOperands.push_back(linalgOp.getOutputOperand(0));
+  }
 
-static bool computeTransposeInfo(LinalgOp linalgOp) {
-  return isTransposeLinalgOp(linalgOp);
+  return transposeOperands;
 }
 
 static bool computeReductionInfo(LinalgOp linalgOp) {
   return linalgOp.getNumReductionLoops() > 1;
 }
 
+static bool computeDynamicInfo(LinalgOp linalgOp) {
+  return linalgOp.hasDynamicShape();
+}
+
+static bool computeTwoOrThreeLoopsInfo(LinalgOp linalgOp) {
+  return linalgOp.getNumParallelLoops() >= 2 &&
+         linalgOp.getNumParallelLoops() <= 3;
+}
+
+static bool computeGenericInfo(LinalgOp linalgOp) {
+  return isa<GenericOp>(linalgOp);
+}
```

Comment:
Would it make more sense to do something like:
```
if (linalgOp.hasDynamicShape() || hasTwoOrThreeLoopsInfo(linalgOp) || isa<GenericOp>(linalgOp))
  return; // or bail out

LinalgOpInfo opInfo (linalgOp);
```

and keep these checks/traits out of LinalgOpInfo? That information is easy to get from the linalgOp itself. I think it's good if we use linalgOp directly when possible instead of adding a wrapper around a simple check in LinalgOpInfo. Otherwise, LinalgOpInfo will grow a lot and will duplicate queries that are already in LinalgOp. WDYT?

I think we should also remove the reductionTrait. I can take care of that later.

Reply:
Done. Though I left the `isReduction` trait. I think that is general enough that it will be useful around but you can remove it if you choose.